theano/gof/fg.py

```diff
@@ -320,13 +321,15 @@ def __remove_clients__(self, r, clients_to_remove,
                               if output.clients or output in self.outputs]
             # If the apply node is not used and is not an output
             if not used_or_output:
-                self.apply_nodes.remove(apply_node)
```

Comment:
@nouiz you mean this line is redundant, right ? Even i feel the same, not sure why added that ```if``` statement.
https://github.com/Theano/Theano/pull/4013/files#diff-8850ac78ebd71ffd0376fa6c58a80a69R324